### PR TITLE
Save correct rotation 315

### LINKED ISSUE #315: Save image and recordings in the correct rotation setting
The images should be saved in the rotation setting in which they were captured

### PR DESCRIPTION
**Pull request by Gmartinica**
# Assignee Section

## Description
Added correct rotation according to the one shown at the GUI to screenshots and recordings of streams. Passed the rotation value from the javascript to the python. Used ffmpeg filters transpose and vflip to rotate the streams. They work altering a stream already created.

- Added ffmpeg-python library
- Divided filename variable in stream_capture.py in two different variables to avoid confusion with image and video filename
- Created rotation function using transpose and vflip filters
- Made the video filename global to use it in the rotation function as the number of seconds had to equal

Comment: Need to `pip install -r requirements.txt` on this branch before testing for this feature to work.

### Steps for Testing
Run

1. `rosgui`
2. `startgui`
3. `roslaunch robot/util/rosRoverStart/local_rover.launch`
4. Visit `localhost:5000` and follow the steps below

#### Test cases with 1 camera

This can be tested on any page with at least one camera panel.

The location of the saved images/videos will be logged in the GUI console.

1. Turn on stream, rotate image, take a screenshot, verify saved image is correctly rotated.
2. Turn on stream, rotate image, start recording a video, stop recording, verify saved video is correctly rotated.
3. Repeat step 1. and 2. for all 4 orientations.

#### Test cases with 2 cameras

This needs to be tested on the multiple streams page.

1. Open 2 camera feeds: Start recording feed1 first, then start recording feed2
2. Stop recording feed2 first, then stop recording feed1
3. Verify saved videos are saved correctly rotated, repeat for different orientations

Note that the video saved will be saved only in the initial orientation, so flipping the orientation during recording will not affect the saved video.

closes #315

The approval from all software team leads is necessary before merging.

# Reviewer Section

Aside from local testing and the General Integration Test it is implied that static analysis should be included in the verification process.

- [x] Local Test Performed Successfully
- [ ] [General Integration Test](https://docs.google.com/document/d/1ug0CpA1cIzURP8DDFSvCt2CEJJSwJ6Ta6B1LG_hYk6I/edit) Performed Successfully

For Pull Requests that do not include code changes, it is not required to perform the tests above.

### CODE REVIEW

File: robot/basestation/app.py [hunk 1]
```diff
@@ -170,8 +170,9 @@ def rover_drive():
 
 @app.route("/capture_image/", methods=["POST", "GET"])
 def capture_image():
-    stream_url = request.args['stream_url']
-    success, message = stream_capture(stream_url)
+    stream_url= request.args['stream_url']
+    rotation = request.args['camera_rotation']
+    success, message = stream_capture(stream_url, rotation)
```
Comment (MartensCedric) [inline robot/basestation/app.py:175]: This isn't your fault because nobody has done it so far, but it would be a good time to document our functions. I suggest you document the methods. It is not clear what the "rotation" argument is. Is it 0,1,2,3, is it 0, 90, 180, 270 or is it "right", "left". Documenting the method would make this clear.

File: robot/basestation/app.py [hunk 2]
```diff
@@ -186,8 +187,9 @@ def initiate_feed_recording():
 @app.route("/stop_feed_recording/", methods=["POST", "GET"])
 def stop_feed_recording():
     stream_url = request.args['stream_url']
+    rotation = request.args['camera_rotation']
     if is_recording_stream(stream_url):
-        success, message = stop_recording_feed(stream_url)
+        success, message = stop_recording_feed(stream_url, rotation)
```
Comment (MartensCedric) [inline robot/basestation/app.py:192]: from a software engineering perspective, I think that it is not the purpose of "stop_recording_feed" to be fed a rotation. I would remove the argument and pass it in the hashmap

File: robot/basestation/stream_capture.py [hunk 3]
```diff
@@ -32,6 +33,7 @@ def start_recording_feed(stream_url):
     """
 
     global active_recordings
+    global video_filename
```
Comment (MartensCedric) [inline robot/basestation/stream_capture.py:36]: Don't forget multiple videos can be recorded at the same time. You should not have a single global video file name. "active_recordings" contains all the video file names already so this is not useful

Comment (MartensCedric) [reply robot/basestation/stream_capture.py:36]: Additionally you should modify "active_recordings" to contain a tuple (name, rotation). Or you could make a class and it would become a hashmap of that class. If you dont understand we can speak of it on Google Hangouts

File: robot/basestation/stream_capture.py [hunk 4]
```diff
@@ -64,6 +66,9 @@ def stop_recording_feed(stream_url):
     success = True
     try:
         proc_video[stream_url].communicate(b'q')
+        if camera_rotation != '0':
```
Comment (MartensCedric) [inline robot/basestation/stream_capture.py:69]: You should really store the rotation as an integer. Cast it to an int worst case but you must handle exceptions so the server doesnt crash

Comment (MartensCedric) [reply robot/basestation/stream_capture.py:69]: I also suggest you don't do this check but just pass in zero

File: robot/basestation/stream_capture.py [hunk 5]
```diff
@@ -91,25 +119,28 @@ def start_ffmpeg_record(stream_url, filename):
...
-    filename =  get_stream_shortname(stream_url) + "_" + formatted_date + ".jpg"
-    filename = image_directory + filename
+    image_filename =  get_stream_shortname(stream_url) + "_" + formatted_date + ".jpg"
+    image_filename = image_directory + image_filename
+
+    error, output = run_shell("ffmpeg -i " + stream_url + " -ss 00:00:01.500 -f image2 -vframes 1 " + image_filename)
 
-    error, output = run_shell("ffmpeg -i " + stream_url + " -ss 00:00:01.500 -f image2 -vframes 1 " + filename)
+    if camera_rotation != '0':
+        rotate_stream(image_filename, camera_rotation)
```
Comment (MartensCedric) [inline robot/basestation/stream_capture.py:141]: same

**Comment by MartensCedric** [comment]
Very cool! There's a few suggestions for better practices. There's also the fact that I don't think your suggestion will always work if multiple people have recordings at the same time since you have a single global filename

**Comment by Gmartinica** [comment]
I did the changes you requested.
I added better documentation for the functions
The rotation value was cast to an int
For the active recordings, I made it into a dictionary. Each key is the stream shortname and the value is a tuple with two elements: The video filename and the rotation.
I pass the value to the python in the hashmap and the function to pass the rotation handles errors too.

**Comment by MartensCedric** [comment]
@Gmartinica 

Just so you know 

>roscore
roslaunch robot/util/rosRoverStart/local_rover.launch
rosgui
startgui

Can be reduced to

```
rosgui
startgui
roslaunch robot/util/rosRoverStart/local_rover.launch
```

(rosgui starts roscore)

**Comment by MartensCedric** [comment]
I tested the branch @Gmartinica and it works pictures and videos are well saved depending on the rotation. I have tested with only 1 webcam. @PeterGhimself will test and ensure it works with more than one webcam. (Peter use the multiple stream page).

File: robot/basestation/stream_capture.py [hunk 6]
```diff
@@ -72,6 +77,61 @@ def stop_recording_feed(stream_url):
...
+    """
+    Rotate stream so that it looks like the rotation of the GUI
+    The rotation value as seen on the GUI is defined as the following:
+    0: No rotation
+    1: 90 degrees clockwise rotation
+    2: 180 degrees rotation
+    3: 90 degrees counterclockwise rotation
+    """
+
+    rotation = int(rotation)
```
Comment (MartensCedric) [inline robot/basestation/stream_capture.py:110]: I would recommend you pass the rotation as an integer instead of having the function 

`rotate_stream(string filename, string rotation)` it makes more sense you have it as
`rotate_stream(string filename, int rotation)`

File: robot/basestation/stream_capture.py [hunk 7]
```diff
@@ -91,25 +151,30 @@ def start_ffmpeg_record(stream_url, filename):
...
 
-def stream_capture(stream_url):
-    """ Given a stream, captures an image.
+def stream_capture(stream_url, camera_rotation):
+    """ Given a stream, captures an image and rotates it as shown in GUI
 
         stream_url : The URL of the stream to capture the image.
+        camera_rotation : The rotation value (0,1,2,3) of the stream as shown
+        in the GUI. 0 = no rotation. 1 = clockwise 90 degrees
+        2 = 180 degrees. 3 = counterclockwise 90 degrees
```
Comment (MartensCedric) [inline robot/basestation/stream_capture.py:160]: You might want to format it just like you did for the other function (by making it multiple lines) for the sake of consistency

**Comment by MartensCedric** [comment]
Hey @Gmartinica You need to resolve some conflicts. Do you know how to do merges?

**Comment by Gmartinica** [comment]
I thought I had solved it when committing, but I just solved the merge conflict

**Comment by MartensCedric** [comment]
> It didnt work, I think you broke something in your last commits
> 
> ```
>   File "/home/cedric/Programming/robotics-prototype/robot/basestation/app.py", line 182, in capture_image
>     success, message = stream_capture(stream_url, rotation)
>   File "/home/cedric/Programming/robotics-prototype/robot/basestation/stream_capture.py", line 176, in stream_capture
>     rotate_stream(image_filename, camera_rotation)
>   File "/home/cedric/Programming/robotics-prototype/robot/basestation/stream_capture.py", line 126, in rotate_stream
>     raise ValueError('Rotation value is not valid')
> ValueError: Rotation value is not valid
> ```

I hadnt even rotated the image lol

File: robot/basestation/app.py [hunk 8]
```diff
@@ -173,6 +173,7 @@ def rover_drive():
 def capture_image():
     stream_url= request.args['stream_url']
     rotation = request.args['camera_rotation']
+    rotation = int(rotation)
```
Comment (MartensCedric) [inline robot/basestation/app.py]:
```suggestion
    rotation = int(request.args['camera_rotation'])
```

File: robot/basestation/app.py [hunk 9]
```diff
@@ -189,6 +190,7 @@ def initiate_feed_recording():
 def stop_feed_recording():
     stream_url = request.args['stream_url']
     rotation = request.args['camera_rotation']
+    rotation = int(rotation)
```
Comment (MartensCedric) [inline robot/basestation/app.py]:
```suggestion
    rotation = int(request.args['camera_rotation'])
```

**Comment by NinjaWeeaboo** [comment]
I look over the code changes quickly and tried breaking it, everything worked perfectly. Good job fellow gamer 😎

File: robot/basestation/app.py [hunk 10]
```diff
@@ -185,13 +187,15 @@ def initiate_feed_recording():
...
+        success, message = start_recording_feed(stream_url)
         return jsonify(success=success, msg=message)
 
 @app.route("/stop_feed_recording/", methods=["POST", "GET"])
 def stop_feed_recording():
     stream_url = request.args['stream_url']
+    rotation = int(request.args['camera_rotation'])
     if is_recording_stream(stream_url):
+        success, message = add_rotation(stream_url, rotation)
         success, message = stop_recording_feed(stream_url)
```
Comment (PeterGhimself) [inline robot/basestation/app.py:199]:
```suggestion
        success, message = add_rotation(stream_url, rotation)
        
        if not success:
            print('add_rotation method failed:', message)
            
        success, message = stop_recording_feed(stream_url)
        
        if not success:
            print('stop_recording_feed method failed:', message)
```

Comment (PeterGhimself) [reply robot/basestation/app.py:199]: let's make use of the potential error messages here, otherwise we're just overwriting those variables without using them at all (aside from sending the final message back to GUI)

File: robot/basestation/stream_capture.py [hunk 11]
```diff
@@ -41,14 +42,14 @@ def start_recording_feed(stream_url):
...
-    filename = save_video_dir + '/' + filename
+    video_filename = get_stream_shortname(stream_url) + '_' + formatted_date + ".mp4"
+    video_filename = save_video_dir + '/' + video_filename
 
     if stream_is_connected:
         print("Started recording " + stream_shortname)
-        active_recordings.append(stream_shortname)
-        threading.Thread(target = start_ffmpeg_record, args = (stream_url, filename)).start() 
-        return True, "Successfully started recording " + stream_shortname + " stream at " + os.path.abspath(filename)
+        active_recordings[stream_shortname] = [video_filename]
```
Comment (PeterGhimself) [inline robot/basestation/stream_capture.py:50]: Is it necessary to assign it as a array? In other words can we remove the square brackets on the right hand side and it would still work? If so let's do that.

File: robot/basestation/stream_capture.py [hunk 12]
```diff
@@ -8,15 +8,16 @@
 from robot.util.utils import run_shell
 import robot.basestation.ros_utils as ros_utils
 from robot.basestation.ros_utils import fetch_ros_master_ip
+import ffmpeg
 
 global proc_video
```
Comment (PeterGhimself) [inline robot/basestation/stream_capture.py:13]: I know you didn't create this variable but could you rename it to be more clear? Something like `video_process`

Comment (PeterGhimself) [reply robot/basestation/stream_capture.py:13]: Another minor issue is that this here is not the correct usage of `global` keyword. AFAIK in most other languages that I have experienced this is the normal approach, but in Python the way it should work is that we declare variables normally on the outter most scope (without the `global` keyword) and then in inner scopes (so like in function definitions) we assert `global` before the variable so that it's known that we are referring to the global variable. We use the keyword in local scopes for both read/writing to the variable. Here's a [reference](https://www.programiz.com/python-programming/global-keyword).

Basically, you'd have to move this line (13) into the top of the functions that make use of this variable.

File: robot/basestation/stream_capture.py [hunk 13]
```diff
@@ -72,6 +77,59 @@ def stop_recording_feed(stream_url):
     print(message)
     return success, message
 
+def add_rotation(stream_url,rotation):
+    """
+    Add rotation value to the value of the key (stream_url) in active recordings
+    Converts the list to a tuple and handles errors
+    """
+    global active_recordings
```
Comment (PeterGhimself) [inline robot/basestation/stream_capture.py:85]: just to follow up on the previous comment about the use of global, this here is an example of the proper way

File: robot/basestation/stream_capture.py [hunk 14]
```diff
@@ -72,6 +77,59 @@ def stop_recording_feed(stream_url):
...
+
+
+def rotate_stream(filename, rotation):
+    """
+    Rotate stream so that it looks like the rotation of the GUI
+    The rotation value as seen on the GUI is defined as the following:
+    0: No rotation
+    1: 90 degrees clockwise rotation
+    2: 180 degrees rotation
+    3: 90 degrees counterclockwise rotation
```
Comment (PeterGhimself) [inline robot/basestation/stream_capture.py:107]: epic :sunglasses:

File: robot/basestation/app.py (hunk 10 above)
Comment (Gmartinica) [reply robot/basestation/app.py:199]: True, I am adding that

File: robot/basestation/stream_capture.py (hunk 11 above)
Comment (Gmartinica) [reply robot/basestation/stream_capture.py:50]: You know what, I was texting some explanation of why when I realized how bad design was that lol. I am changing that. I will not use the list

File: robot/basestation/stream_capture.py (hunk 12 above)
Comment (Gmartinica) [reply robot/basestation/stream_capture.py:13]: Yeah, I remember reading that in Slack. I am changing that

**Comment by MartensCedric** [comment]
I still have this issue 
image


@AFknUsername Do you have this problem?

**Comment by MartensCedric** [comment]
image

**Comment by MartensCedric** [comment]
image

**Comment by MartensCedric** [comment]
Its fishy that it names something `.mp4.mp4`

**Comment by AFknUsername** [comment]
@MartensCedric @Gmartinica 
I also get the same error as @MartensCedric

**Comment by AFknUsername** [comment]
Screenshot from 2020-06-13 16-39-08

**Comment by AFknUsername** [comment]
After restarting everything it seems to be working fine. 
Rotated stream multiple times and took multiple videos and screenshots. Also switched cameras a few times.

**Comment by MartensCedric** [comment]
`    stream = ffmpeg.input(filename)
AttributeError: module 'ffmpeg' has no attribute 'input'
`

**Comment by MartensCedric** [comment]
image

**Comment by MartensCedric** [comment]
I didnt have python-ffmpeg but installed it just in case but it seems it doesnt fix the issue I have
